### ISSUE TITLE
HTML elements not written properly in Jetbrains IDE

### ISSUE THREAD

**Issue by gntsketches** [dictation-toolbox/Caster#541]
Hi, in Jetbrains Webstorm, html elements often are written with additional partial tags. For example, when I say "division", I get this `<div></div>div></div>` (where I just expected `<div></div>`) ... note the second opening div is missing the `<`. The curser is placed after the "i" in the first closing </div>. "Paragraph" will be written as `<p>/n</p>p>` with the cursor at the newline. This behavior is inconsistent depending on the surrounding context. 

By contrast, things seem to work fine in Notepad++

**Comment by LexiconCode** [comment]
This is most likely an issue related to auto complete being triggered in Webstorm.

**Comment by LexiconCode** [comment]
is this true of all commands in Webstorm?

**Comment by LexiconCode** [comment]
https://github.com/dictation-toolbox/Caster/issues/56
@kendonB looks like that is still relevant, Thoughts?

ContextAction would work well for exceptions but this is pretty systemic within the HTML grammar for webstorm.

Maybe the grammar could be Re-factored so it does not trigger auto complete in webstorm.

**Comment by LexiconCode** [comment]
I confirm that this is definitely related to auto complete. Webstorm not only automatically closes HTML tags but moves the cursor as well producing the issue.

 - The following works as expected
`"division": R(Text("<division>"), rdescript="HTML: Division"),`
Outputs
`<division></division>`

- Whereas the original
"division":  R(Text("<div></div>") + Key("left/10:6"), rdescript="HTML: Division"),
Outputs
<div></div>div></div>

**Comment by gntsketches** [comment]
Oh, of course.

I definitely use the autocomplete functionality alot when coding, and continuingly/more it would seem with voice. If it's possible to not autocomplete for HTML code in particular that might be helpful. But disabling it globally wouldn't be good.

**Comment by LexiconCode** [comment]
We deftly don't want to disable auto complete. Preferably I wouldn't want anyone to have to touch their settings to make this work. There are a few suggestions in the above referenced issue that need to be investigated as an alternative.

**Comment by gntsketches** [comment]
Would it be possible to prefix/suffix the command with "open" or "close" and just get part of the tag? That way the autocomplete functionality wouldn't conflict. 

It would also be useful in those times when stray element tags need to be dealt with. This could also be useful for (), [], {}... it would be cool to be able to say "brax closed" in similar situation. 

Apologies if something like this case already exists, if so I may have missed it since I'm new.

**Comment by kendonB** [comment]
I think as a general rule, we should aim to support the out-of-the-box behavior of the apps we cover. Echoing this comment: https://github.com/dictation-toolbox/Caster/issues/56#issuecomment-489483568, I actually think this particular issue is a bug with the editor as you should be able to type over auto completions. I would raise the issue with the maintainers of the editor you using. In the short term we should address this in caster too.

I read somewhere that using paste instead of Text/Key is unreliable; I'm fairly sure that it's slower so I don't think that's a very good option. @alexboche has been using the clipboard for a recent PR so might be able to comment on the relative merits. 

If we don't address this using paste, then I think we just have to have a `ContextAction` where we need it. 

@gntsketches this is a "good first issue". Would you like to have a go at it following the approach here: https://github.com/dictation-toolbox/Caster/pull/493? I think you would just need to edit `castervoice\lib\ccr\html\html.py` in the `develop` branch.

**Comment by LexiconCode** [comment]
Testing out the paste action seems to work fine
`from dragonfly.actions.action_paste import Paste as Text`

Now if only something like this would work to change the import based on context.
```
ContextAction(default=<main>
</main>actions=[
                (AppContext(executable=["webstorm64", "webstorm"]), from dragonfly.actions.action_paste import Paste as Text),
                ])),
```

**Comment by LexiconCode** [comment]
What you think of this workaround that contextually changes the imports in the HTML.py ccr @kendonB @gntsketches?

**Comment by gntsketches** [comment]
Thanks for the encouragement kendonB. In general I am just getting started here and I actually have zero experience coding Python. Hopefully this is the sort of thing that I would be able to attempt as I get more familiar with everything. 

LexiconCode I will also attempt to try out your new variation sometime before too long, with the same caveat that I just gave kendonB. In the short term, I was able to just turn of auto-completion of HTML in Webstorm :)
